### PR TITLE
Fix dataset deletion/recreation across processes

### PR DESCRIPTION
**Pull request by benjaminpkane**
Dataset singletons are currently keyed by name. The consequence of this is that when a dataset is deleted and recreated in one process, any other processes will raise a dataset is deleted exception if the original dataset singleton is still in memory. This causes the App server to remain in a persisted error state if the session is attached to the original (now deleted) dataset.

This is a sensitive change, but references to dataset singleton instances seem minimal, and keying by an immutable value seems more robust.

The below now works with this change.
```py
import fiftyone as fo
import fiftyone.zoo as foz

dataset, session = fo.quickstart()
# wait for app to launch

# delete and recreate quickstart
dataset.delete()
dataset = foz.load_zoo_dataset("quickstart")

# refreshing the App page loads the new quickstart, currently broken on `release/v0.22.2`
```

Screenshot of existing error
Screenshot 2023-10-11 at 10 29 40 AM

### CODE REVIEW

File: fiftyone/core/singletons.py [hunk 1]
```diff
@@ -24,7 +26,13 @@ def __new__(metacls, *args, **kwargs):
         return cls
 
     def __call__(cls, name=None, _create=True, *args, **kwargs):
-        instance = cls._instances.pop(name, None)
+        instance = None
+        if name:
+            doc = foo.get_db_conn().datasets.find_one(
```
Comment (brimoor) [inline fiftyone/core/singletons.py:31]: Just calling out that this adds another db connection into the process of loading a dataset. Not impactful for direct mongo connections but would add more overhead to API connections.

**Review by brimoor — COMMENTED** [review]
Just brainstorming- is there a way to detect and handle deletion/recreation without changing how singletons are keyed?

It's tricky because the database *has* to be involved somewhere; the singleton cache is simply outdated. I guess the only hope is to detect this rare case only upon failure, not checking a priori for it.

**Comment by benjaminpkane** [comment]
> Just brainstorming- is there a way to detect and handle deletion/recreation without changing how singletons are keyed?
> 
> It's tricky because the database _has_ to be involved somewhere; the singleton cache is simply outdated. I guess the only hope is to detect this rare case only upon failure, not checking a priori for it.

Yeah, I can try catching the `instance._update_last_loaded_at()` attempt. That is where the exception occurs

File: fiftyone/core/singletons.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benjaminpkane) [reply fiftyone/core/singletons.py:31]: Yes, it's certainly a low-usage code path, but I will try Brian's suggestion

**Comment by benjaminpkane** [comment]
Ok comments addressed. The test added is a best attempt to simulate a the multiprocessing with orphaning mongodb collections/documents

File: fiftyone/core/singletons.py [hunk 2]
```diff
@@ -36,7 +36,11 @@ def __call__(cls, name=None, _create=True, *args, **kwargs):
             instance.__init__(name=name, _create=_create, *args, **kwargs)
             name = instance.name  # `__init__` may have changed `name`
         else:
-            instance._update_last_loaded_at()
+            try:
+                instance._update_last_loaded_at()
+            except ValueError:
+                instance._deleted = True
+                return cls.__call__(name=name, _create=False, *args, **kwargs)
```
Comment (brimoor) [inline fiftyone/core/singletons.py:43]: nit: `_create=_create` might be slightly more readable to convey that you're just re-invoking the method?

**Review by brimoor — APPROVED** [review]
LGTM, nice solution to this! 💪

**Review by brimoor — APPROVED** [review]
LGTM

**Pull request merged**